appinventor/buildserver/src/com/google/appinventor/buildserver/stats/SimpleStatReporter.java

```diff
@@ -0,0 +1,129 @@
+  private static final Logger LOG = Logger.getLogger(SimpleStatReporter.class.getCanonicalName());
+
+  /**
+   * BuildStats encapsulate all of the measurements of a particular build, including the
+   * timing measurements of all of the build's subprocesses.
+   */
+  public static class BuildStats {
+    private final long start = System.currentTimeMillis();
+    private final Map<String, Long> stages = new LinkedHashMap<>();
+    private long end;
+    private long duration;
+    private long last = start;
+
+    private BuildStats() {
+      // Not instantiable outside this class
+    }
+
+    public long getStart() {
+      return start;
+    }
+
+    public long getEnd() {
+      return end;
+    }
+
+    public long getDuration() {
+      return duration;
+    }
+
+    public Map<String, Long> getStages() {
+      return Collections.unmodifiableMap(stages);
+    }
+  }
+
+  private final Map<Compiler, BuildStats> activeBuilds = new HashMap<>();
+
+  private final Deque<BuildStats> successfulBuilds = new LinkedList<>();
+  private final Deque<BuildStats> failedBuilds = new LinkedList<>();
+  private final Deque<BuildStats> orderedBuilds = new LinkedList<>();
+
+  @Override
+  public void startBuild(Compiler compiler) {
+    activeBuilds.put(compiler, new BuildStats());
+  }
+
+  @Override
+  public void nextStage(Compiler compiler, String stage) {
+    BuildStats stats;
+    synchronized (this) {
+      stats = activeBuilds.get(compiler);
+    }
+    if (stats == null) {
+      LOG.warning("Got compiler with uninitialized stats object");
+      return;
+    }
+    stats.stages.put(stage, System.currentTimeMillis() - stats.last);
+    stats.last = System.currentTimeMillis();
+  }
+
+  @Override
+  public void stopBuild(Compiler compiler, boolean success) {
+    BuildStats stats;
+    synchronized (this) {
+      stats = activeBuilds.remove(compiler);
+    }
+    if (stats == null) {
+      LOG.warning("Got compiler with uninitialized stats object");
+      return;
+    }
+    stats.end = System.currentTimeMillis();
+    stats.duration = stats.end - stats.start;
+    queueAndExpire(stats, success ? successfulBuilds : failedBuilds);
+    queueAndExpire(stats, orderedBuilds);
+  }
+
+  public Collection<BuildStats> getSuccessStats() {
+    return Collections.unmodifiableCollection(successfulBuilds);
+  }
+
+  public Collection<BuildStats> getFailureStats() {
+    return Collections.unmodifiableCollection(failedBuilds);
+  }
+
+  public Collection<BuildStats> getOrderedStats() {
+    return Collections.unmodifiableCollection(orderedBuilds);
+  }
+
+  @SuppressWarnings("SynchronizationOnLocalVariableOrMethodParameter")
+  private void queueAndExpire(BuildStats stats, final Deque<BuildStats> target) {
+    synchronized (target) {
+      if (target.size() == MAX_STATS) {
+        target.pollFirst();
+      }
+      target.push(stats);
+    }
+    synchronized (orderedBuilds) {
+      if (target.size() == MAX_STATS) {
```

Comment:
@ewpatton What is this extra code synchronized under orderedBuilds doing here. It looks like you call queueAndExpire seperately for successful (or failed) builds and then for orderedBuilds. Am I missing something?

Reply:
Yes, you're right.  I've removed the second synchronized block.